### DISCUSSION TITLE
Huan Yang GT series vector controlled VFD

### DISCUSSION THREAD

**Discussion by Meliusja** [grblHAL/plugins#6] [Q&A] [answered]
Hi,
I'm adding a V3 vfd type for my VFD which is an authentic Huan Yang GT series which works in V/Hz mode or vector control mode.  I've noticed some odd things in some of the other modbus implementations like the v2 seems to be 50Hz specific.  And I would like to fix this in the GT series VFD implementation.  This vfd has 100% modbus compatibility and I can communicate with the generic libmodbus from a PC.  My issue is that the VFD  takes a speed reference as a percentage of its programmed top speed and in order to determine the correct value to send from an RPM value I would need to read three parameters.

P0.03 Maximum reference frequency 
P2.02 Motor rated frequency
P2.03 Motor rated speed

P2.02 and p2.03 can be read in one read since they are consecutive.  But presently the plugin is designed for a single calibration read.

This would be simpler if the plugin received the speed setpoint as a percentage of the grbl programmed maximum and then I assumed the vfd max speed was equal.

Any thoughts?

Jeff

[GT-series-manual.pdf](https://github.com/grblHAL/plugins/files/8721237/GT-series-manual.pdf)

**Comment by terjeio** [comment, marked as answer]
> P2.02 and p2.03 can be read in one read since they are consecutive. But presently the plugin is designed for a single calibration read.

That could be changed as needed.

> This would be simpler if the plugin received the speed setpoint as a percentage of the grbl programmed maximum

You mean the configured maximum (from the settings)?

> and then I assumed the vfd max speed was equal.

The percentage is can be calclulated in the plugin so IMO there is no need to do that in the core. Also, the max RPM used in the core is defaulted from the settings and can be overridden by the plugin at run time by writing it to the `hal.spindle.rpm_max` variable (when the spindle is selected and active).

**Reply by Meliusja** [reply]
Thank you very much.  I completed testing and everything works well.   I noticed that if the vfd is not there when GRBL starts it never seems to call config method again to complete the initialization.  I copied the structure of V1 and V2 exactly so I am not sure where I went wrong.  Should the config method get called on a grblhal reset?

Is there any interest in including this support for the huanyang GT series VFD?
Jeff
